src/main/groovy/org/pih/warehouse/inventory/CycleCountTransactionCreator.groovy

```diff
@@ -0,0 +1,128 @@
+package org.pih.warehouse.inventory
+
+import grails.validation.ValidationException
+import org.springframework.stereotype.Component
+
+import org.pih.warehouse.core.Constants
+import org.pih.warehouse.product.Product
+
+/**
+ * Component responsible for creating and persisting the Transactions that result from a cycle count.
+ */
+@Component
+class CycleCountTransactionCreator {
+
+    ProductAvailabilityService productAvailabilityService
+    ProductInventorySnapshotTaker productInventorySnapshotTaker
+    TransactionIdentifierService transactionIdentifierService
+
+    /**
+     * Given a completed cycle count, will create and persist the resulting product inventory snapshot transaction,
+     * as well as the credit/debit adjustment transaction if there are discrepancies.
+     *
+     * @param itemQuantityOnHandIsUpToDate
+     */
+    List<Transaction> createTransactions(CycleCount cycleCount, boolean itemQuantityOnHandIsUpToDate=false) {
+        List<Transaction> transactions = []
+        Date transactionDate = new Date()
+
+        Transaction productInventoryTransaction = createProductInventoryTransaction(cycleCount, transactionDate)
+        transactions.add(productInventoryTransaction)
+
+        Transaction adjustmentTransaction = createAdjustmentTransaction(
+                cycleCount, transactionDate, itemQuantityOnHandIsUpToDate)
+        if (adjustmentTransaction) {
+            transactions.add(adjustmentTransaction)
+        }
+
+        return transactions
+    }
+
+    private Transaction createAdjustmentTransaction(
+            CycleCount cycleCount, Date transactionDate, boolean itemQuantityOnHandIsUpToDate) {
+        // We only care about the cycle count items from the most recent count. We create the entries first to avoid
+        // needlessly creating the transaction itself if there are no discrepancies.
+        List<TransactionEntry> entries = []
+        for (CycleCountItem cycleCountItem : cycleCount.itemsOfMostRecentCount) {
+            if (cycleCountItem.quantityCounted == null) {
+                continue
+            }
+
+            // We need to compare the quantity counted against the most up to date QoH in product availability.
```

Comment:
> We need to compare the quantity counted against the most up to date QoH in product availability.

A side question popped up in my head, and I do not remember if or when this was discussed. If something is already allocated, then should this be counted as well? So we are always counting QoH? Is there a chance that we miss in the count something that is allocated, but not yet sent?

Reply:
a good question. Lets discuss with Manon